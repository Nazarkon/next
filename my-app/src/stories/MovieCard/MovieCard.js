"use client"
import React, { useState } from 'react';
import PropTypes from 'prop-types';
import DotsMenu from '../DotsMenu/DotsMenu';
import './MovieCard.scss';
import { createMovieItemsList } from '../../helpers/MovieItemsHelpers';
import { redirect } from 'next/navigation';

const MovieCard = ({ id, imageUrl, name, year, genreList, navigationList }) => {
  const [imageLoadError, setImageLoadError] = useState(false);


  const handleImageError = () => {
    setImageLoadError(true);
  };

  const handleMovieRoute = () => {
    const searchParams = new URLSearchParams(document.location.search);
    const sortBy = searchParams.get('sortBy');
    const genre = searchParams.get('genre');
    redirect(`/movie-details/${id}?sortBy=${sortBy}&genre=${genre}`);
  };

  return (
    <div className="movie-card-container">
      <div className="movie-card-container-image">
        <div className="movie-card-container-menu">
          {navigationList && (
            <DotsMenu
              navigationList={navigationList}
              movieID={id}
              className="movie-card-container-menu"
            />
          )}
        </div>
        <>
          {imageLoadError ? (
            <div
              onClick={handleMovieRoute}
              className="movie-card-error"
              alt="error"
              src={'../assets/images/404-error.png'}
            />
          ) : (
            <img
              onClick={handleMovieRoute}
              className="movie-card-image"
              alt="film poster"
              src={imageUrl}
              onError={handleImageError}
            />
          )}
        </>
      </div>
      <div className="movie-card-container-title">
        <h4 className="movie-card-title">{name}</h4>
        <p className="movie-card-year">{year}</p>
      </div>
      <span className="movie-card-genres">{createMovieItemsList(genreList)}</span>
    </div>
  );
};

MovieCard.propTypes = {
  id: PropTypes.number.isRequired,
  imageUrl: PropTypes.string.isRequired,
  name: PropTypes.string.isRequired,
  year: PropTypes.string.isRequired,
  genreList: PropTypes.array.isRequired,
  handleClick: PropTypes.func,
  navigationList: PropTypes.array
};

export default MovieCard;
