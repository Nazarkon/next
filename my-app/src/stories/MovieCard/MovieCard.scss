.movie-card-container{
    width: 100%;
    cursor: pointer;
    max-width: 332px;
    display: flex;
    justify-content: center;
    flex-direction: column;
      .movie-card-container-image{
        position: relative;
        .movie-card-image{
            width: 100%;
            height: 100%;
            min-height: 500px;
        }
        .movie-card-container-menu{
          position: absolute;
          right: 5px;
          top: 10px;
        }
        .movie-card-error{
          width: 100%;
          height: 100%;
          display: flex;
          justify-content: center;
          min-height: 500px;
          background-image: url('../assets/images/404-error.png');
          background-size: contain;
          background-repeat: no-repeat;
        }
      }
      .movie-card-container-title{
        display: flex;
        justify-content: space-between;
        align-items: center;
            .movie-card-title{
              color: #FFFFFF;
              font-size: 18px;
              flex: 1;
            }
            .movie-card-year{
                border: 1px solid #FFFFFF;
                border-radius: 4px;
                padding: 4px 8px;
                font-size: 18px;
                color: #FFFFFF;
            }
      }
        .movie-card-genres{
          color: #FFFFFF;
        }
}
.movie-card-container-more-button{
  width: 50px;
  height: 50px;
  border-radius: 50%;
  position: absolute;
  display: flex;
  justify-content: center;
  align-items: center;
  flex-direction: column;
  gap: 5px;
  top: 10px;
  right: 10px;
  z-index: 10;
  background: rgba(42, 32, 45, 1);
    .movie-card-container-more-button-dot{
      background-color: white;
      width: 5px;
      height: 5px;
      border-radius: 50%;
    }
}