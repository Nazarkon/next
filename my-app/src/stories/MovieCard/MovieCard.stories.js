import React from 'react';

import MovieCard from './MovieCard';

import filmImage from '../assets/images/Bitmap.png';

export default {
  title: 'Components/MovieCard',
  component: MovieCard,
  argTypes: {
    imageUrl: {
      control: 'text',
      description: 'Image url that should be a string',
      defaultValue: filmImage
    },
    name: {
      control: 'text',
      description: 'Film title',
      defaultValue: 'Pulp Fiction'
    },
    year: {
      control: 'text',
      description: 'Should be a year when film was presented',
      defaultValue: '2003'
    },
    genreList: {
      control: 'array',
      describe: 'List of genres that connected to this film',
      defaultValue: ['Action', 'Adventure']
    },
    handleClick: {
      action: 'clicked'
    }
  }
};

const Template = (args) => (
    <MovieCard {...args} />
);

export const Default = Template.bind({});
