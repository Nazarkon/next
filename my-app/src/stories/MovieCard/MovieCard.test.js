import React from 'react';
import { render } from '@testing-library/react';
import '@testing-library/jest-dom';
import MovieCard from './MovieCard';
import { BrowserRouter } from 'react-router-dom';

describe('MovieCard component', () => {
  const mockData = {
    image: 'Bitmap.png',
    name: 'Pulp Fiction',
    year: '2003',
    genreList: ['Action', 'Adventure']
  };

  it('matches snapshot', () => {
    const { container } = render(
      <BrowserRouter>
        <MovieCard
          id={123}
          imageUrl={mockData.image}
          name={mockData.name}
          year={mockData.year}
          genreList={mockData.genreList}
        />
      </BrowserRouter>
    );
    expect(container).toMatchSnapshot();
  });
  it('Check that component renders movie information correctly', () => {
    const { getByAltText, getByText } = render(
      <BrowserRouter>
        <MovieCard
          id={123}
          imageUrl={mockData.image}
          name={mockData.name}
          year={mockData.year}
          genreList={mockData.genreList}
        />
      </BrowserRouter>
    );

    expect(getByAltText('film poster')).toHaveAttribute('src', 'Bitmap.png');

    expect(getByText('Pulp Fiction')).toBeInTheDocument();

    expect(getByText('2003')).toBeInTheDocument();

    expect(getByText('Action&Adventure')).toBeInTheDocument();
  });
});
