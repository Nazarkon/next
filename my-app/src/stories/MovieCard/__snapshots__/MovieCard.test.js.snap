// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieCard component matches snapshot 1`] = `
<div>
  <div
    class="movie-card-container"
  >
    <a
      href="/movie-details/123"
    >
      <div
        class="movie-card-container-image"
      >
        <img
          alt="film poster"
          class="movie-card-image"
          src="Bitmap.png"
        />
      </div>
    </a>
    <div
      class="movie-card-container-title"
    >
      <h4
        class="movie-card-title"
      >
        Pulp Fiction
      </h4>
      <p
        class="movie-card-year"
      >
        2003
      </p>
    </div>
    <span
      class="movie-card-genres"
    >
      Action&Adventure
    </span>
  </div>
</div>
`;
